### ISSUE TITLE
Bug: .safeParse() should not throw

### ISSUE THREAD

**Issue by kirkwaiblinger** [colinhacks/zod#3756]
From the docs, 

> If you don't want Zod to throw errors when validation fails, use .safeParse. This method returns an object containing either the successfully parsed data or a ZodError instance containing detailed information about the validation problems.

This works fine and well so long as you stick to built-in refinements and schemas. But, if you make your own transformations/refinements, and they throw, then safeparse will throw.
```ts
const throwingSchema = z.any().transform((args, context) => {
   throw new Error('lol');
});

const resultShouldntThrow = throwingSchema.safeParse("please don't throw!");
```

I understand that user-defined `.transform()` and `.refine()` functions are not intended to throw, so they constitute an invalid schema. However, I would expect that to be reported from `.safeParse()` in a non-throwing way, for example

```ts
throwingSchema.safeParse("please don't throw!"); // { valid: false, code: 'INVALID_SCHEMA', message: 'Exception was thrown during user-defined `.transform()` callback' }
```

If we really want a method called "safeX" to throw, I would expect it to come in the form of an _opt-in_ option, such as
```ts
throwingSchema.safeParse("please don't throw!", { throwOnInvalidSchema: true }); // Uncaught: ZodErrorInvalidSchema
```

Otherwise, we're stuck with
```ts
let parseResult;
try {
   parseResult = throwingSchema.safeParse("please don't throw!");
} catch (error) {
   // handle secondary error path
}
if (!parseResult.valid) {
   // handle primary error path
}
```
which defeats the purpose of `safeParse` in the first place.

**Comment by kirkwaiblinger** [comment]
Relevant

**Comment by danawoodman** [comment]
Yeah I just came across this today and honestly its incredibly frustrating to have a method that claims to be safe but can fail in unexpected ways. This should be fixed or the method should be renamed to `sortaSafeParse`

**Comment by danawoodman** [comment]
dear dosubot, i hate you
